source/slang/slang-compiler.cpp

```diff
@@ -1244,6 +1243,13 @@ SlangResult dissassembleDXILUsingDXC(
             {
                 preprocessorDefinitions.Add(define.Key, define.Value);
             }
+            {
```

Comment:
The blame here is on earlier changes of mine, not this PR, but it would be great it we had a better abstraction over these cases where there are nested "scopes" on which options can be set. There is a clear pattern here where for a given compilation option we want to merge/reduce whatever was specified on the linkage, compile request, and then translation unit. For an option like preprocessor defines or include paths this would mean concatenating all the values, but for something like a simple toggle it would mean taking the value from the "inner-most" thing that specified that flag.

This is a somewhat subtle bit of technical debt, and it isn't something I've been willing to build a larger "system" around just yet, because the other alternative (just making each setting "live" at one exact level of the hierarchy) is much more appealing when it is possible.

Reply:
I should note there there is a small issue here that the original code here has a subtle design choice where the ordering is:

1. compile request `#define`s
2. translation unit `#define`s

Because both are being combined into the same dictionary, the fact that the translation unit comes second means it can "override" a define from the compile request (e.g., if the compile request has `-D FOO=1` while the translation unit has `-D FOO=2`, the final dictionary should have `FOO=2`.

Putting the linkage last here is then backwards, because the linkage is the "outer-most" scope for state, and its values should be overridden by either of the other two.

Reply:
One more note: the above comment has me realizing this code is probably actually buggy because we don't want to use `Dictionary::Add(...)` here given that it asserts on a key that already exists.

We probably either want to reverse the order of translation unit and compile request (work from "inner" scope to "outer") and use `AddIfNotExists` so that we keep the inner-most settings once they are added, *or* to put the linkage first (work from "outer" to "inner") and just use `Dictionary::operator[]` to set the defines in a way that clobbers existing entries.

Aside: I'm not sure I like the design choice in `Dictionary` of asserting on a pre-existing key. It bites us an awful lot and it doesn't always seem to be catching real bugs when it does.

Reply:
Ok - makes sense and will fix after merge. Agree on the Dictionary. We might want to to rename/refactor the Dictionary somewhat anyway.